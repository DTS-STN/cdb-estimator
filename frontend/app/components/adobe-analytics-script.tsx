import * as adobeAnalytics from '~/utils/adobe-analytics-utils';

export function AdobeAnalyticsHeadScript() {
  const { ADOBE_ANALYTICS_JQUERY_SRC, ADOBE_ANALYTICS_SRC } = globalThis.__appEnvironment;

  if (adobeAnalytics.isEnabled()) {
    return (
      <>
        <script src={ADOBE_ANALYTICS_JQUERY_SRC} />
        <script src={ADOBE_ANALYTICS_SRC} />
      </>
    );
  }
  return <></>;
}

export function AdobeAnalyticsBottomScript() {
  if (adobeAnalytics.isEnabled() && adobeAnalytics.isDebug()) {
    return <script id="gc-analytics-bottom-script">_satellite.setDebug(true);</script>;
  }
  return <></>;
}
