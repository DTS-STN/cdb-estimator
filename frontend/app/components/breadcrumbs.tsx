import type { ReactNode } from 'react';

import type { Params, Path, To } from 'react-router';

import { faChevronRight } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import type { ResourceKey, TFunction } from 'i18next';
import { useTranslation } from 'react-i18next';

import { InlineLink } from './links';

import { useLanguage } from '~/hooks/use-language';
import type { I18nRouteFile } from '~/i18n-routes';
import * as adobeAnalytics from '~/utils/adobe-analytics-utils';

export interface BreadcrumbRouteLink {
  file: I18nRouteFile;
  hash?: Path['hash'];
  lang?: Language;
  params?: Params;
  search?: Path['search'];
}

export interface BreadcrumbItem {
  labelKey?: ResourceKey;
  label?: { fr: string; en: string };
  destinationRoute?: BreadcrumbRouteLink;
  destination?: { en: To; fr: To };
}

export interface BreadcrumbsProps {
  className?: string;
  items: BreadcrumbItem[];
}

function getLabel(item: BreadcrumbItem, t: TFunction<['common'], undefined>, currentLanguage?: Language) {
  if (item.labelKey !== undefined) {
    return t(item.labelKey);
  }
  return currentLanguage === 'fr' ? item.label?.fr : item.label?.en;
}

function getDestination(item: BreadcrumbItem, currentLanguage?: Language) {
  if (item.destinationRoute !== undefined) {
    return item.destinationRoute;
  }
  return currentLanguage === 'fr' ? item.destination?.fr : item.destination?.en;
}

export function Breadcrumbs({ className, items }: BreadcrumbsProps) {
  const { t } = useTranslation(['common']);
  const { currentLanguage } = useLanguage();

  return (
    <nav id="wb-bc" className={className} property="breadcrumb" aria-labelledby="breadcrumbs">
      <h2 id="breadcrumbs" className="sr-only">
        {t('common:breadcrumbs.you-are-here')}
      </h2>
      <div className="container">
        <ol className="flex flex-wrap items-center gap-x-3 gap-y-1" typeof="BreadcrumbList">
          {items.map((item, idx) => {
            const label = getLabel(item, t, currentLanguage);
            const destination = getDestination(item, currentLanguage);
            return (
              <li key={label} property="itemListElement" typeof="ListItem" className="flex items-center">
                {idx !== 0 && <FontAwesomeIcon icon={faChevronRight} className="mr-2 size-3 text-slate-700" />}
                <Breadcrumb
                  destination={destination}
                  data-gc-analytics-customclick={adobeAnalytics.getCustomClick(`Breadcrumb:${idx} link`)}
                >
                  {label}
                </Breadcrumb>
              </li>
            );
          })}
        </ol>
      </div>
    </nav>
  );
}

function isBreadcrumbRouteLink(destination: To | BreadcrumbRouteLink): destination is BreadcrumbRouteLink {
  return typeof destination === 'object' && 'file' in destination;
}

function Breadcrumb({ children, destination, ...rest }: { children: ReactNode; destination?: To | BreadcrumbRouteLink }) {
  if (!destination) {
    return <span property="name">{children}</span>;
  }

  return isBreadcrumbRouteLink(destination) ? (
    <InlineLink
      file={destination.file}
      hash={destination.hash}
      params={destination.params}
      search={destination.search}
      property="item"
      typeof="WebPage"
      {...rest}
    >
      <span property="name">{children}</span>
    </InlineLink>
  ) : (
    <InlineLink to={destination} property="item" typeof="WebPage" {...rest}>
      <span property="name">{children}</span>
    </InlineLink>
  );
}
