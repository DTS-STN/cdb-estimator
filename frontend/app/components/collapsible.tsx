import { useId } from 'react';
import type { ComponentProps, ReactNode } from 'react';

import { cn } from '~/utils/tailwind-utils';

export type CollapsibleSummaryProps = ComponentProps<'summary'>;

export function CollapsibleSummary({ children, className, ...props }: CollapsibleSummaryProps) {
  return (
    <summary className={cn('marker:cursor-pointer hover:text-blue-800', className)} {...props}>
      <div className="ml-2 inline-block text-lg font-medium text-slate-950 underline hover:cursor-pointer">{children}</div>
    </summary>
  );
}

export interface CollapsibleProps extends ComponentProps<'details'> {
  contentClassName?: string;
  summary: ReactNode;
}

export function Collapsible({ children, contentClassName, id, summary, ...props }: CollapsibleProps) {
  const uniqueId = useId();
  const summaryId = `${id ?? uniqueId}-summary`;
  const contentId = `${id ?? uniqueId}-content`;
  return (
    <details id={id ?? uniqueId} {...props} className="my-2">
      <CollapsibleSummary id={summaryId}>{summary}</CollapsibleSummary>
      <div id={contentId} className={cn('my-2 border-l-[2px] border-slate-950 px-6 text-lg', contentClassName)}>
        {children}
      </div>
    </details>
  );
}
