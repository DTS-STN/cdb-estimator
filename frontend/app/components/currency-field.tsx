import { useTranslation } from 'react-i18next';
import { NumericFormat } from 'react-number-format';

import type { InputFieldProps } from './input-field';
import { InputField } from './input-field';

import { removeNumericFormatting } from '~/utils/string-utils';
import { cn } from '~/utils/tailwind-utils';

export interface CurrencyFieldProps extends Omit<InputFieldProps, 'type' | 'value' | 'defaultValue'> {
  allowNegative?: boolean;
  value?: string | number | null;
  defaultValue?: string | number | null;
}

export function CurrencyField({ allowNegative = false, maxLength, defaultValue, ...rest }: CurrencyFieldProps) {
  const { i18n } = useTranslation(['common']);
  const strippedDefaultValue = Number(removeNumericFormatting(defaultValue?.toString()));
  const val = !Number.isInteger(strippedDefaultValue) ? strippedDefaultValue.toFixed(2) : strippedDefaultValue;

  return (
    <NumericFormat
      customInput={InputField}
      thousandSeparator={i18n.language === 'fr' ? ' ' : ','}
      allowedDecimalSeparators={i18n.language === 'fr' ? [',', '.'] : ['.']}
      allowNegative={allowNegative}
      decimalScale={2}
      decimalSeparator={i18n.language === 'fr' ? ',' : '.'}
      fixedDecimalScale={false}
      thousandsGroupStyle={'thousand'}
      maxLength={maxLength ?? (i18n.language === 'en' ? 15 : 16)}
      type="text"
      beforeInput={i18n.language === 'en' ? <span className="mr-1">$</span> : undefined}
      afterInput={i18n.language === 'fr' ? <span className="ml-1">$</span> : undefined}
      className={cn('inline', rest.className)}
      defaultValue={val}
      {...rest}
    />
  );
}
