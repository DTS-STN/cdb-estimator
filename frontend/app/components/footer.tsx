import { useTranslation } from 'react-i18next';

import { cn } from '~/utils/tailwind-utils';

export function Footer() {
  const { t } = useTranslation(['common']);

  return (
    <footer id="footer" className="bg-[#26374A]">
      <section>
        <h2 className="sr-only" id="accessibleSectionHeader1">
          {t('common:footer.header-accessible')}
        </h2>
        <h3 className="container pt-[24px] pb-2 text-[19px] leading-[21px] font-bold text-white">{t('footer.header')}</h3>
        <div className="bg-[url('/footer_bg_img.svg')] bg-clip-border bg-bottom bg-no-repeat sm:bg-right-bottom">
          <nav role="navigation" aria-labelledby="accessibleSectionHeader1">
            <ul className="container flex flex-col pb-[22px] sm:grid sm:grid-cols-3">
              {FooterItem(
                t('common:footer.all-contacts.text'),
                t('common:footer.all-contacts.href'),
                'footerLine mb-3 relative pb-[26px]',
              )}

              {FooterItem(t('common:footer.department-and-agencies.text'), t('common:footer.department-and-agencies.href'))}
              {FooterItem(t('common:footer.about-government.text'), t('common:footer.about-government.href'))}
              {FooterItem(t('common:footer.jobs.text'), t('common:footer.jobs.href'))}
              {FooterItem(t('common:footer.taxes.text'), t('common:footer.taxes.href'))}
              {FooterItem(t('common:footer.canada-and-world.text'), t('common:footer.canada-and-world.href'))}
              {FooterItem(
                t('common:footer.immigration-and-citizenship.text'),
                t('common:footer.immigration-and-citizenship.href'),
              )}
              {FooterItem(
                t('common:footer.environment-and-natural-resources.text'),
                t('common:footer.environment-and-natural-resources.href'),
              )}
              {FooterItem(t('common:footer.money-and-finance.text'), t('common:footer.money-and-finance.href'))}
              {FooterItem(t('common:footer.travel-and-tourism.text'), t('common:footer.travel-and-tourism.href'))}
              {FooterItem(
                t('common:footer.national-security-and-defence.text'),
                t('common:footer.national-security-and-defence.href'),
              )}
              {FooterItem(t('common:footer.science-and-innovation.text'), t('common:footer.science-and-innovation.href'))}
              {FooterItem(t('common:footer.business.text'), t('common:footer.business.href'))}
              {FooterItem(
                t('common:footer.culture-history-and-sports.text'),
                t('common:footer.culture-history-and-sports.href'),
              )}
              {FooterItem(t('common:footer.indigenous-peoples.text'), t('common:footer.indigenous-peoples.href'))}
              {FooterItem(t('common:footer.benefits.text'), t('common:footer.benefits.href'))}
              {FooterItem(
                t('common:footer.policing-justice-and-emergencies.text'),
                t('common:footer.policing-justice-and-emergencies.href'),
              )}
              {FooterItem(t('common:footer.veterans-and-military.text'), t('common:footer.veterans-and-military.href'))}
              {FooterItem(t('common:footer.health.text'), t('common:footer.health.href'))}
              {FooterItem(
                t('common:footer.transport-and-infrastructure.text'),
                t('common:footer.transport-and-infrastructure.href'),
              )}
              {FooterItem(t('common:footer.youth.text'), t('common:footer.youth.href'))}
            </ul>
          </nav>
        </div>
      </section>
      <section className="bg-[#F8F8F8]">
        <div className="container flex min-h-[96px] justify-between">
          <div className="flex flex-row items-center">
            <nav role="navigation" aria-labelledby="accessibleSectionHeader2">
              <h2 className="sr-only" id="accessibleSectionHeader2">
                {t('common:footer.about-site')}
              </h2>
              <ul className="flex flex-col py-4 whitespace-nowrap md:flex-row">
                {SubFooterItem(t('common:footer.social-media.text'), t('common:footer.social-media.href'), 'first')}
                {SubFooterItem(t('common:footer.mobile-applications.text'), t('common:footer.mobile-applications.href'))}
                {SubFooterItem(t('common:footer.about-canada-ca.text'), t('common:footer.about-canada-ca.href'))}
                {SubFooterItem(t('common:footer.terms-and-conditions.text'), t('common:footer.terms-and-conditions.href'))}
                {SubFooterItem(t('common:footer.privacy.text'), t('common:footer.privacy.href'))}
              </ul>
            </nav>
          </div>
          <div className="mr-[5px] flex min-h-[96px] shrink-0 items-end md:items-center">
            <img
              className="my-[15px] h-[25px] w-[105px] md:h-[40px] md:w-[164px]"
              src="/wmms-blk.svg"
              alt={t('common:footer.gc-symbol')}
            />
          </div>
        </div>
      </section>
    </footer>
  );
}

function FooterItem(text: string, href: string, className: string | undefined = undefined) {
  return (
    <li className={cn('my-1 w-64 list-none sm:w-56 lg:w-80', className)}>
      <a className="text-[14px] leading-[19px] font-[400] !text-white" href={href}>
        {text}
      </a>
    </li>
  );
}

function SubFooterItem(text: string, href: string, type: 'first' | 'subsequent' = 'subsequent') {
  return (
    <li className={cn('my-1 pr-4', type !== 'first' ? "md:before:mr-3 md:before:content-['â€¢']" : undefined)}>
      <a className="text-[14px] leading-[19px] font-[400] text-[#284162]" href={href}>
        {text}
      </a>
    </li>
  );
}
