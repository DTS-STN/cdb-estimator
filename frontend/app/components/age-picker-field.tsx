import { Fragment, useState } from 'react';
import type { ChangeEvent, JSX, ReactNode } from 'react';

import { useTranslation } from 'react-i18next';

import { InputError } from '~/components/input-error';
import { InputHelp } from '~/components/input-help';
import { InputLabel } from '~/components/input-label';
import { InputLegend } from '~/components/input-legend';
import { useLanguage } from '~/hooks/use-language';
import { calculateAge } from '~/utils/age-utils';
import { getLocalizedMonths } from '~/utils/date-utils';
import { cn } from '~/utils/tailwind-utils';

type DatePart = 'year' | 'month';

type AriaErrorMessage = Readonly<Record<DatePart, string | undefined>>;

/**
 * Order of date parts based on language
 */
const DATE_PART_ORDER = {
  en: ['month', 'year'] as const,
  fr: ['month', 'year'] as const,
} as const satisfies Record<Language, DatePart[]>;

const inputStyles = {
  base: 'rounded-lg border-gray-500 focus:border-blue-500 focus:ring-blue-500 block focus:ring focus:outline-none',
  disabled: 'disabled:bg-gray-100 disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-70',
  error: 'border-red-500 focus:border-red-500 focus:ring-red-500',
} as const;

/**
 * Props for the AgePickerField component
 */
export interface AgePickerFieldProps {
  defaultValues?: {
    year?: number;
    month?: number;
  };
  disabled?: boolean;
  displayAge?: boolean;
  errorMessages?: {
    all?: string;
    month?: string;
    year?: string;
  };
  helpMessagePrimary?: React.ReactNode;
  helpMessagePrimaryClassName?: string;
  helpMessageSecondary?: React.ReactNode;
  helpMessageSecondaryClassName?: string;
  id: string;
  legend: ReactNode;
  names: {
    month: string;
    year: string;
  };
  required?: boolean;
}

/**
 * AgePickerField component
 *
 * @param props - Props for the AgePickerField component
 * @returns JSX.Element
 */
export const AgePickerField = ({
  defaultValues,
  disabled,
  displayAge,
  errorMessages,
  helpMessagePrimary,
  helpMessagePrimaryClassName,
  helpMessageSecondary,
  helpMessageSecondaryClassName,
  id,
  legend,
  names,
  required,
}: AgePickerFieldProps): JSX.Element => {
  const { t } = useTranslation(['common', 'estimator']);
  const { currentLanguage = 'en' } = useLanguage(); // english by default
  const currentDatePartOrder = DATE_PART_ORDER[currentLanguage];

  // Generate unique IDs for accessibility
  const baseId = `age-picker-${id}`;
  const ids = {
    wrapper: baseId,
    legend: `${baseId}-legend`,
    error: {
      all: `${baseId}-error-all`,
      month: `${baseId}-error-month`,
      year: `${baseId}-error-year`,
    },
    help: {
      primary: `${baseId}-help-primary`,
      secondary: `${baseId}-help-secondary`,
    },
  };

  // Combine IDs for aria-describedby attribute
  const ariaDescribedBy: string = [
    ids.legend,
    helpMessagePrimary ? ids.help.primary : false,
    helpMessageSecondary ? ids.help.secondary : false,
  ]
    .filter(Boolean)
    .join(' ');

  // Generate error messages for each date part
  const ariaErrorMessage: AriaErrorMessage = currentDatePartOrder.reduce((acc, datePart) => {
    const errors = [errorMessages?.all ? ids.error.all : false, errorMessages?.[datePart] ? ids.error[datePart] : false]
      .filter(Boolean)
      .join(' ');

    return { ...acc, [datePart]: errors || undefined };
  }, {} as AriaErrorMessage);

  const [month, setMonth] = useState(defaultValues?.month);
  const [year, setYear] = useState(defaultValues?.year);
  const age = month && year && calculateAge(month, year);
  const ageAdditionalInfo = (() => {
    if (age !== undefined) {
      if (age >= 65) return t('estimator:age.max-eligible-age-info');
      else if (age < 18) return t('estimator:age.min-eligible-age-info');
    }
    return undefined;
  })();

  const handleMonthChange = (event: ChangeEvent<HTMLSelectElement>) => {
    setMonth(parseInt(event.target.value));
  };

  const handleYearChange = (event: ChangeEvent<HTMLInputElement>) => {
    setYear(parseInt(event.target.value));
  };

  // Define age picker part fields
  const agePickerPartFields = {
    year: (
      <AgePickerYearField
        id={id}
        defaultValue={year}
        name={names.year}
        label={t('common:age-picker.year.label')}
        className="w-full sm:w-32"
        ariaDescribedBy={ariaDescribedBy}
        ariaErrorMessage={ariaErrorMessage.year}
        onChange={handleYearChange}
        required={required}
        disabled={disabled}
      />
    ),
    month: (
      <AgePickerMonthField
        id={id}
        defaultValue={month}
        name={names.month}
        label={t('common:age-picker.month.label')}
        placeholder={t('common:age-picker.month.placeholder')}
        className="w-full sm:w-auto"
        currentLanguage={currentLanguage}
        ariaDescribedBy={ariaDescribedBy}
        ariaErrorMessage={ariaErrorMessage.month}
        onChange={handleMonthChange}
        required={required}
        disabled={disabled}
      />
    ),
  } as const satisfies Record<DatePart, JSX.Element>;

  return (
    <div id={ids.wrapper}>
      <fieldset className="space-y-2">
        <InputLegend id={ids.legend} required={required}>
          {legend}
        </InputLegend>

        {/* Error Messages */}
        {errorMessages && Object.values(errorMessages).filter(Boolean).length > 0 && (
          <div className="space-y-2">
            {errorMessages.all && <InputError id={ids.error.all}>{errorMessages.all}</InputError>}
            {currentDatePartOrder
              .filter((datePart) => Boolean(errorMessages[datePart]))
              .map((datePart) => (
                <p key={datePart}>
                  <InputError id={ids.error[datePart]}>{errorMessages[datePart]}</InputError>
                </p>
              ))}
          </div>
        )}

        {/* Help Messages - Primary */}
        {helpMessagePrimary && (
          <InputHelp id={ids.help.primary} className={helpMessagePrimaryClassName}>
            {helpMessagePrimary}
          </InputHelp>
        )}

        {/* Date Picker Part Fields */}
        <div className="flex flex-col space-y-2 sm:flex-row sm:space-y-0 sm:space-x-2">
          {currentDatePartOrder.map((datePart) => (
            <Fragment key={datePart}>{agePickerPartFields[datePart]}</Fragment>
          ))}

          {displayAge === true && age !== undefined && year !== undefined && year > 1000 && year < 9999 && (
            <AgeDisplay age={age} additionalInfo={ageAdditionalInfo} />
          )}
        </div>

        {/* Help Messages - Secondary */}
        {helpMessageSecondary && (
          <InputHelp id={ids.help.secondary} className={helpMessageSecondaryClassName}>
            {helpMessageSecondary}
          </InputHelp>
        )}
      </fieldset>
    </div>
  );
};

/**
 * Props for the AgePickerMonthField component
 */
interface AgePickerMonthFieldProps {
  ariaDescribedBy: string;
  ariaErrorMessage?: string;
  className?: string;
  currentLanguage?: Language;
  defaultValue?: number;
  disabled?: boolean;
  id: string;
  label: string;
  name: string;
  onChange: (event: ChangeEvent<HTMLSelectElement>) => void;
  placeholder: string;
  required?: boolean;
}

/**
 * AgePickerMonthField component
 *
 * @param props - Props for the AgePickerMonthField component
 * @returns JSX.Element
 */
function AgePickerMonthField({
  ariaDescribedBy,
  ariaErrorMessage,
  className,
  currentLanguage,
  defaultValue,
  disabled,
  id,
  label,
  name,
  onChange,
  placeholder,
  required,
}: AgePickerMonthFieldProps): JSX.Element {
  const baseId = `age-picker-${id}-month`;
  const ids = {
    wrapper: baseId,
    label: `${baseId}-label`,
    select: `${baseId}-select`,
    optionUnselected: `${baseId}-option-unselected`,
    option: (monthIndex: number) => `${baseId}-option-${monthIndex}`,
  };
  const hasErrorMessage = !!ariaErrorMessage;
  const months = getLocalizedMonths(currentLanguage ?? 'en');

  return (
    <div id={ids.wrapper} className="space-y-2">
      <InputLabel id={ids.label} htmlFor={ids.select}>
        {label}
      </InputLabel>
      <select
        aria-describedby={ariaDescribedBy}
        aria-errormessage={ariaErrorMessage}
        aria-invalid={hasErrorMessage}
        aria-labelledby={ids.label}
        aria-required={required}
        className={cn(inputStyles.base, inputStyles.disabled, hasErrorMessage && inputStyles.error, className)}
        defaultValue={defaultValue}
        disabled={disabled}
        id={ids.select}
        name={name}
        onChange={onChange}
        required={required}
      >
        <option id={ids.optionUnselected} disabled hidden selected={defaultValue === undefined}>
          {placeholder}
        </option>
        {months.map((month) => {
          return (
            <option id={ids.option(month.value)} key={month.value} value={month.value}>
              {month.text}
            </option>
          );
        })}
      </select>
    </div>
  );
}

/**
 * Props for the AgePickerYearField component
 */
interface AgePickerYearFieldProps {
  ariaDescribedBy: string;
  ariaErrorMessage?: string;
  className?: string;
  defaultValue?: number;
  disabled?: boolean;
  id: string;
  label: string;
  name: string;
  required?: boolean;
  onChange: (event: ChangeEvent<HTMLInputElement>) => void;
}

/**
 * AgePickerYearField component
 *
 * @param props - Props for the AgePickerYearField component
 * @returns JSX.Element
 */
function AgePickerYearField({
  ariaDescribedBy,
  ariaErrorMessage,
  className,
  defaultValue,
  disabled,
  id,
  label,
  name,
  onChange,
  required,
}: AgePickerYearFieldProps): JSX.Element {
  const baseId = `age-picker-${id}-year`;
  const ids = {
    wrapper: baseId,
    label: `${baseId}-label`,
    input: `${baseId}-input`,
  };
  const hasErrorMessage = !!ariaErrorMessage;

  return (
    <div id={ids.wrapper} className="space-y-2">
      <InputLabel id={ids.label} htmlFor={ids.input}>
        {label}
      </InputLabel>
      <input
        aria-describedby={ariaDescribedBy}
        aria-errormessage={ariaErrorMessage}
        aria-invalid={hasErrorMessage}
        aria-labelledby={ids.label}
        aria-required={required}
        className={cn(inputStyles.base, inputStyles.disabled, hasErrorMessage && inputStyles.error, className)}
        defaultValue={defaultValue}
        disabled={disabled}
        id={ids.input}
        min={1900}
        name={name}
        required={required}
        type="number"
        inputMode="numeric"
        pattern="[0-9]*"
        maxLength={4}
        onChange={onChange}
      />
    </div>
  );
}

/**
 * Props for the AgeDisplay component
 */
interface AgeDisplayProps {
  age: number;
  additionalInfo?: string;
}

/**
 * AgeDisplay component
 *
 * @param props - Props for the AgeDisplay component
 * @returns JSX.Element
 */
function AgeDisplay({ age, additionalInfo }: AgeDisplayProps): JSX.Element {
  const { t } = useTranslation(['common']);

  return (
    <div role="alert" aria-live="assertive" className="space-y-1.5 sm:ml-30">
      <label className="block">
        <span className="font-semibold">{t('age-picker.your-age.label')}</span>
      </label>

      <div className="block max-w-prose">{`${age} ${t(age === 1 ? 'age-picker.your-age.value-suffix-singular' : 'age-picker.your-age.value-suffix-plural')}`}</div>

      <div>{additionalInfo}</div>
    </div>
  );
}
