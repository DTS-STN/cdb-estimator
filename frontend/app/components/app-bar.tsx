import { Menu } from '~/components/menu';

interface AppBarProps {
  children: React.ReactNode;
}

export function AppBar({ children }: AppBarProps) {
  return (
    <div className="bg-slate-700">
      <div className="align-center container mx-auto flex flex-wrap justify-between">
        <div className="align-center flex">
          <Menu>{children}</Menu>
        </div>
      </div>
    </div>
  );
}
