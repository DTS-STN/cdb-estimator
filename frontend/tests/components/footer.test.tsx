import { render } from '@testing-library/react';
import { useTranslation } from 'react-i18next';
import { afterEach, describe, expect, it, vi } from 'vitest';

import { Footer } from '~/components/footer';

describe('Footer', () => {
  afterEach(() => {
    vi.mocked(useTranslation).mockReset();
  });

  it('should render footer', () => {
    const { container } = render(<Footer />);
    expect(container).toMatchSnapshot('expected html');
  });
});
