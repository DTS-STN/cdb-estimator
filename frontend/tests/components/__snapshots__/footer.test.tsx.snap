// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Footer > should render footer > expected html 1`] = `
<div>
  <footer
    class="bg-[#26374A]"
    id="footer"
  >
    <section>
      <h2
        class="sr-only"
        id="accessibleSectionHeader1"
      >
        common:footer.header-accessible
      </h2>
      <h3
        class="container pt-[24px] pb-2 text-[19px] leading-[21px] font-bold text-white"
      >
        footer.header
      </h3>
      <div
        class="bg-[url('/footer_bg_img.svg')] bg-clip-border bg-bottom bg-no-repeat sm:bg-right-bottom"
      >
        <nav
          aria-labelledby="accessibleSectionHeader1"
          role="navigation"
        >
          <ul
            class="container flex flex-col pb-[22px] sm:grid sm:grid-cols-3"
          >
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80 footerLine mb-3 relative pb-[26px]"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.all-contacts.href"
              >
                common:footer.all-contacts.text
              </a>
            </li>
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.department-and-agencies.href"
              >
                common:footer.department-and-agencies.text
              </a>
            </li>
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.about-government.href"
              >
                common:footer.about-government.text
              </a>
            </li>
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.jobs.href"
              >
                common:footer.jobs.text
              </a>
            </li>
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.taxes.href"
              >
                common:footer.taxes.text
              </a>
            </li>
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.canada-and-world.href"
              >
                common:footer.canada-and-world.text
              </a>
            </li>
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.immigration-and-citizenship.href"
              >
                common:footer.immigration-and-citizenship.text
              </a>
            </li>
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.environment-and-natural-resources.href"
              >
                common:footer.environment-and-natural-resources.text
              </a>
            </li>
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.money-and-finance.href"
              >
                common:footer.money-and-finance.text
              </a>
            </li>
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.travel-and-tourism.href"
              >
                common:footer.travel-and-tourism.text
              </a>
            </li>
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.national-security-and-defence.href"
              >
                common:footer.national-security-and-defence.text
              </a>
            </li>
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.science-and-innovation.href"
              >
                common:footer.science-and-innovation.text
              </a>
            </li>
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.business.href"
              >
                common:footer.business.text
              </a>
            </li>
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.culture-history-and-sports.href"
              >
                common:footer.culture-history-and-sports.text
              </a>
            </li>
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.indigenous-peoples.href"
              >
                common:footer.indigenous-peoples.text
              </a>
            </li>
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.benefits.href"
              >
                common:footer.benefits.text
              </a>
            </li>
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.policing-justice-and-emergencies.href"
              >
                common:footer.policing-justice-and-emergencies.text
              </a>
            </li>
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.veterans-and-military.href"
              >
                common:footer.veterans-and-military.text
              </a>
            </li>
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.health.href"
              >
                common:footer.health.text
              </a>
            </li>
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.transport-and-infrastructure.href"
              >
                common:footer.transport-and-infrastructure.text
              </a>
            </li>
            <li
              class="my-1 w-64 list-none sm:w-56 lg:w-80"
            >
              <a
                class="text-[14px] leading-[19px] font-[400] !text-white"
                href="common:footer.youth.href"
              >
                common:footer.youth.text
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </section>
    <section
      class="bg-[#F8F8F8]"
    >
      <div
        class="container flex min-h-[96px] justify-between"
      >
        <div
          class="flex flex-row items-center"
        >
          <nav
            aria-labelledby="accessibleSectionHeader2"
            role="navigation"
          >
            <h2
              class="sr-only"
              id="accessibleSectionHeader2"
            >
              common:footer.about-site
            </h2>
            <ul
              class="flex flex-col py-4 whitespace-nowrap md:flex-row"
            >
              <li
                class="my-1 pr-4"
              >
                <a
                  class="text-[14px] leading-[19px] font-[400] text-[#284162]"
                  href="common:footer.social-media.href"
                >
                  common:footer.social-media.text
                </a>
              </li>
              <li
                class="my-1 pr-4 md:before:mr-3 md:before:content-['•']"
              >
                <a
                  class="text-[14px] leading-[19px] font-[400] text-[#284162]"
                  href="common:footer.mobile-applications.href"
                >
                  common:footer.mobile-applications.text
                </a>
              </li>
              <li
                class="my-1 pr-4 md:before:mr-3 md:before:content-['•']"
              >
                <a
                  class="text-[14px] leading-[19px] font-[400] text-[#284162]"
                  href="common:footer.about-canada-ca.href"
                >
                  common:footer.about-canada-ca.text
                </a>
              </li>
              <li
                class="my-1 pr-4 md:before:mr-3 md:before:content-['•']"
              >
                <a
                  class="text-[14px] leading-[19px] font-[400] text-[#284162]"
                  href="common:footer.terms-and-conditions.href"
                >
                  common:footer.terms-and-conditions.text
                </a>
              </li>
              <li
                class="my-1 pr-4 md:before:mr-3 md:before:content-['•']"
              >
                <a
                  class="text-[14px] leading-[19px] font-[400] text-[#284162]"
                  href="common:footer.privacy.href"
                >
                  common:footer.privacy.text
                </a>
              </li>
            </ul>
          </nav>
        </div>
        <div
          class="mr-[5px] flex min-h-[96px] shrink-0 items-end md:items-center"
        >
          <img
            alt="common:footer.gc-symbol"
            class="my-[15px] h-[25px] w-[105px] md:h-[40px] md:w-[164px]"
            src="/wmms-blk.svg"
          />
        </div>
      </div>
    </section>
  </footer>
</div>
`;
